Implemented composite pattern for filesystem

schlingling opened the pull request:
closes #131

Introduces a clean composite pattern for all filesystem related artefacts like it is done in ADAP.

<img width="1171" alt="image" src="https://user-images.githubusercontent.com/58326464/231539562-6302d17c-3571-4e62-92f9-4d22f08ab831.png">


The naming convention of the .ts-files yields the class-hierachy, so it is easier to understand.

<img width="289" alt="image" src="https://user-images.githubusercontent.com/58326464/231539770-18ecf8e9-9911-4cce-9520-851cdede8349.png">

schlingling commented:
@rhazn FYI, since you are on leave, i requested Felix for an Review :)

rhazn commented:
Thank you, if this is still open next week ping me and I can take a look then as well. If Felix is fine with it that's enough as well of course.

felix-oq commented on libs/execution/src/lib/types/io-types/filesystem.ts:
An idea I have in mind: Introduce a generic type for FileSystem that is used as a type for files. Something like `FileSystem<F extends FileSystemFile>` and in the application probably use it as `FileSystem<BinaryFile>`. What do you think about that?

schlingling replied:
Limits it to just store File of the same type in the FileSystem but state now, this functionality is enough. So i like that.

felix-oq commented on libs/execution/src/lib/types/io-types/filesystem-inmemory.ts:
```diff
@@ -4,94 +4,96 @@
-        currentFileSystemIndex.set(part, childFileSystemIndex);
+  private processPath(path: string): string[] {
+    const parts = path
+      .toLowerCase()
```
Why do you use `toLowerCase` here? That way, users cannot refer to files / folders that have capital letters in their names.

schlingling replied:
@rhazn  once raised the requirement, to handle paths case insensitive, thats why i am using toLowerCase() (see https://github.com/jvalue/jayvee/pull/126#discussion_r1098412123)

rhazn replied:
Hmm I thought that is how unix does it but apparently I was wrong, we can change to care for capitalization, yeah.

schlingling replied:
okay, changed to case sensitive pathing

felix-oq commented on libs/execution/src/lib/types/io-types/filesystem-inmemory.ts:
```diff
@@ -4,94 +4,96 @@
+        continue; // Skip current dirs in path
+      }
+      if (part === InMemoryFileSystem.PARENT_DIR) {
+        processedParts.pop(); // Go level up in folder hierarchy if ..
```
What if `processedParts` is empty at this point? In your current implementation, a path `../folder/file.txt` is semantically equal to `./folder/file.txt` which should not be the case.

schlingling replied:
Ouh thats a good point, thanks for that hint! 

Actually in the current implementation of `InMemoryFileSystem` which always starts at its "virtual" root-dir it does not make sense to call getFile/putFile in general with a relative path (like relativ to what since it is in memory?).

I would mention this in the documentation, if this is okay?

felix-oq replied:
Yes, I'd argue that you should always call it with absolute paths, so you always start in the root directory. Then you should ensure, that the method is only called with absolute paths and error on any paths that ascent beyond the root directory (e.g. `/folder/../..`).

schlingling replied:
is implemented in new version

felix-oq commented on libs/execution/src/lib/types/io-types/filesystem-node-directory.ts:
```diff
@@ -0,0 +1,52 @@
+  }
+
+  add(fileSystemNode: FileSystemNode) {
+    this.children.push(fileSystemNode);
```
You need to pay attention to duplicate names here. E.g. if a user tries to add a file `file.txt` but a file with the same name already exists as a child of this directory.

schlingling replied:
added check if node name exists. This implementation requires now in each node to have unique node names.

Drawback is, that we cant have a dir and a file with same name in same directory, but this can be implemented later on if needed.

felix-oq replied:
I think this should work right away, so the in-memory file system is semantically equivalent to the one of the corresponding ZIP archive. In case we postpone it, we should definitely track it via an issue.

schlingling replied:
Now it works equally, so no track in new issue neccesary.

felix-oq commented on libs/execution/src/lib/types/io-types/filesystem-node-directory.ts:
```diff
@@ -0,0 +1,52 @@
+    return this;
+  }
+
+  findFile(fileName: string): FileSystemNode | undefined {
```
I'd expect a `FileSystemFile` as a return value here.

Also, you should shift this method to `FileSystemNode` and have the different implementations in its subclasses. That way you can avoid `instanceof` checks.

And also note that this method returns a file with the given name that is found first. To find a specific file at a given path, you need the path as a parameter, so it can be descent recursively along the path.

schlingling replied:
i removed that method since it wasnt used.

felix-oq commented on libs/execution/src/lib/types/io-types/filesystem-node-directory.ts:
```diff
@@ -0,0 +1,52 @@
+
+  getChildNode(childName: string): FileSystemNode | undefined {
+    for (const child of this.children) {
+      if (child.name.toLowerCase() === childName) {
```
Also here, I think you should not apply `toLowerCase` to any names so we can distinguish between lower case and capital letters.

schlingling replied:
removed it and changed behavior to support case sensitive paths

felix-oq commented on libs/execution/src/lib/types/io-types/filesystem-inmemory.ts:
```diff
@@ -4,94 +4,96 @@
 
-  // Hierachical file system
-  private fileSystemIndex: Map<string, BinaryFile | FileSystem> = new Map();
+  getFile(path: string): FileSystemFile<unknown> | null {
```
After refactoring the `findFile` method, it can be called here on `rootDirectory` to find the desired file.

schlingling replied:
removed find file, so thats obsolet.

felix-oq commented on libs/execution/src/lib/types/io-types/filesystem-inmemory.ts:
```diff
@@ -4,94 +4,96 @@
-  getFile(filePath: string): BinaryFile | null {
-    const processedParts = this.processPath(filePath);
+  putFile(path: string, file: BinaryFile): FileSystem | null {
+    const processedParts = this.processPath(path);
```
Similar to `findFile`, you could add a method to `FileSystemNode` that covers this functionality. Then, you could simply call `rootDirectory.putFile(path, file)` (or something similar) at this point to descent recursively and add the file at the desired location. This again would avoid the `instanceof` checks.

schlingling replied:
My approach was to keep the FileSystemNode (composite) simple for type saftey (just add, and get methods) and move all the logic to the client in this case the `FileSystem`. 

See [wiki article](https://en.wikipedia.org/wiki/Composite_pattern#:~:text=Design%20for%20type%20safety%3A%20Child%2Drelated%20operations%20are%20defined%20only%20in%20the%20Composite%20class.%20Clients%20must%20treat%20Leaf%20and%20Composite%20objects%20differently.%20But%20type%20safety%20is%20gained%20because%20clients%20cannot%20perform%20child%2Drelated%20operations%20on%20Leaf%20objects.), which is describing design for type saftey

felix-oq replied:
I'd argue for the uniformity approach rather than the type safety approach, since this is emphasized by the composite pattern and leads to more cohesive code in my opinion. Also see [here](https://en.wikipedia.org/wiki/Composite_pattern#:~:text=The%20Composite%20design%20pattern%20emphasizes%20uniformity%20over%20type%20safety.) in the wiki article.

schlingling replied:
Oh yes you are right, convinced me :D I changed with new commit to uniformity approach using clean recursion. I hope you like it :D

felix-oq reviewed (changes requested):
To avoid the `instanceof` checks, consider introducing new methods in `FileSystemNode` that perform the recursive descents. Those can then be called in the `FileSystem` class.

Also I think that paths should generally be treated case-sensitive.

schlingling commented on libs/execution/src/lib/types/io-types/filesystem.ts:
```diff
@@ -15,17 +16,17 @@ export interface FileSystem extends IOTypeImplementation<IOType.FILE_SYSTEM> {
+   * @returns {FileSystem | null} - The directory or null if the file failed to insert
    */
-  putFile(filePath: string, file: BinaryFile): FileSystem;
+  putFile(path: string, file: BinaryFile): FileSystem | null;
```
thanks, changed

schlingling commented on libs/execution/src/lib/types/io-types/filesystem.ts:
```diff
@@ -15,17 +16,17 @@ export interface FileSystem extends IOTypeImplementation<IOType.FILE_SYSTEM> {
+   * @function putNode
+   * @param {string} path - The path to the node.
+   * @param {FileSystemNode} node - The node to save.
+   * @returns {FileSystem | null} - The directory or null if the file failed to insert
```
thanks, changed.

schlingling commented:
> To avoid the `instanceof` checks, consider introducing new methods in `FileSystemNode` that perform the recursive descents. Those can then be called in the `FileSystem` class.
> 
> Also I think that paths should generally be treated case-sensitive.

Appreciate your feedback, thank you! 
- I changed the behavior of caseinsensitivitiy to casesensitive paths.
- I optimized current implementations in `FileSystemInMemory` simplified code in whole composite hierachy and removed unused peaces
- I followed the [composite pattern variant proposed by wikipedia](https://en.wikipedia.org/wiki/Composite_pattern#:~:text=Design%20for%20type%20safety%3A%20Child%2Drelated%20operations%20are%20defined%20only%20in%20the%20Composite%20class.%20Clients%20must%20treat%20Leaf%20and%20Composite%20objects%20differently.%20But%20type%20safety%20is%20gained%20because%20clients%20cannot%20perform%20child%2Drelated%20operations%20on%20Leaf%20objects.) namley "Design for type saftey" so i need to treat composites differently in the client (FileSystem)

felix-oq commented on libs/execution/src/lib/types/io-types/filesystem.ts:
```diff
@@ -13,19 +13,19 @@ import { IOTypeImplementation } from './io-type-implementation';
 export interface FileSystem extends IOTypeImplementation<IOType.FILE_SYSTEM> {
   /**
-   * Retrieves a file from the file system.
+   * Retrieves a file from the file system. Depending on the implementation relative Paths (./) and parent directories (../) are treated differntly.
```
Why might `.` and `..` be treated differently in different implementations? I think they should have fixed semantics (i.e. `.` for current folder, `..` for parent folder), regardless of the concrete implementation.

schlingling replied:
Right, i changed the docs

felix-oq commented on libs/execution/src/lib/types/io-types/filesystem.ts:
```diff
@@ -13,19 +13,19 @@ import { IOTypeImplementation } from './io-type-implementation';
-   * @param {File} file - The file to save.
-   * @returns {FileSystem}
+   * Saves a file to the file system. Depending on the implementation relative Paths (./) and parent directories (../) are treated differntly.
+   * @function putNode
```
```suggestion
   * @function putFile
```

Or omit the line entirely, because the function name is obvious from the function signature below.

felix-oq commented on libs/extensions/std/exec/src/archive-interpreter-executor.ts:
```diff
@@ -83,7 +82,12 @@ export class ArchiveInterpreterExecutor
             content,
           );
-          root.putFile(relPath, file);
+          if (root.putFile(relPath, file) == null) {
+            return R.err({
+              message: `File with name ${file.name} and path ${relPath} failed to insert`,
+              diagnostic: { node: context.getCurrentNode(), property: 'name' },
+            });
+          }
```
Better use `assert` here, because we expect that all `putFile` calls succeed at runtime:
```suggestion
          const addedFile = root.putFile(relPath, file);
          assert(addedFile != null);
```

felix-oq commented on libs/extensions/std/exec/src/file-picker-executor.ts:
```diff
@@ -2,6 +2,8 @@
 //
 // SPDX-License-Identifier: AGPL-3.0-only
 
+import assert = require('assert');
```
```suggestion
import {strict as assert} from 'assert';
```

schlingling commented:
> To avoid the `instanceof` checks, consider introducing new methods in `FileSystemNode` that perform the recursive descents. Those can then be called in the `FileSystem` class.
> 
> Also I think that paths should generally be treated case-sensitive.

all implemented with new commits

felix-oq reviewed (approved):
A few more hints for improving the implementation, but otherwise this is good to go.

schlingling commented on libs/extensions/std/exec/src/archive-interpreter-executor.ts:
```diff
@@ -2,6 +2,7 @@
 //
 // SPDX-License-Identifier: AGPL-3.0-only
 
+import assert = require('assert');
```
changed

schlingling commented on libs/execution/src/lib/types/io-types/filesystem-node.ts:
```diff
@@ -0,0 +1,10 @@
+  constructor(public name: string) {}
+  abstract addChild(fileSystemNode: FileSystemNode): FileSystemNode | null;
+  abstract getNode(path: string): FileSystemNode | null;
+  abstract putNode(path: string, node: FileSystemNode): FileSystemNode | null;
```
changed

schlingling commented on libs/execution/src/lib/types/io-types/filesystem-node-directory.ts:
```diff
@@ -0,0 +1,66 @@
+    return null;
+  }
+
+  override putNode(path: string, node: FileSystemNode): FileSystemNode | null {
```
splitted a little bit. But would prefer, having all necessary information on one view, since it is a recursive function. Added additionally comments to make is easier to read.
